File: internal/engine/atkey.go
package engine

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
)

// userAnnounceEndpoint is Academic Torrents' per-user announce resolution
// endpoint. Called with the operator's API key (uid/pass) as cookies, it
// returns the announce URL that carries that account's passkey, e.g.
// https://academictorrents.com/announce.php?passkey=abc123. This is the same
// mechanism AT's own smartnode tooling uses to attribute hosted torrents to
// an account; see the "Hosted by" attribution section of AT's docs. It's a
// var (not const) so tests can point it at a local stub.
var userAnnounceEndpoint = "https://academictorrents.com/apiv2/userannounce"

// atTrackerHosts are the only tracker hosts keep-at will ever send an
// operator's API key-derived announce URL to. Everything else - third-party
// trackers listed in .torrent files, other AT hosts, http:// (non-https)
// variants - must never see it.
var atTrackerHosts = map[string]bool{
	"academictorrents.com":      true,
	"ipv6.academictorrents.com": true,
}

// atAnnounceURL returns the per-user announce URL to use for a tracker URL,
// or "" if the tracker URL is not one keep-at is allowed to key (i.e. not an
// https URL on an AT tracker host).
func atAnnounceURL(announce string, userAnnounce, userAnnounceIPv6 string) string {
	u, err := url.Parse(announce)
	if err != nil || u.Scheme != "https" {
		return ""
	}
	switch u.Hostname() {
	case "academictorrents.com":
		return userAnnounce
	case "ipv6.academictorrents.com":
		return userAnnounceIPv6
	}
	return ""
}

// resolveUserAnnounce fetches the per-user announce URL for the given API key
// by calling AT's userannounce endpoint with the key sent as cookies (the
// key's format is literally "uid=N;pass=H", which is valid Cookie header
// syntax). Returns the announce URL and its ipv6.academictorrents.com
// variant. The returned URLs contain the account's passkey and must be
// treated as sensitive: never log them, never write them to disk.
func resolveUserAnnounce(ctx context.Context, client *http.Client, apiKey string) (string, string, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, userAnnounceEndpoint, nil)
	if err != nil {
		return "", "", fmt.Errorf("engine: building userannounce request: %w", err)
	}
	req.Header.Set("Cookie", apiKey)

	resp, err := client.Do(req)
	if err != nil {
		return "", "", fmt.Errorf("engine: fetching user announce URL: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(io.LimitReader(resp.Body, 1<<20))
	if err != nil {
		return "", "", fmt.Errorf("engine: reading userannounce response: %w", err)
	}
	if resp.StatusCode != http.StatusOK {
		return "", "", fmt.Errorf("engine: userannounce endpoint returned %s", resp.Status)
	}

	var out struct {
		UserAnnounce string `json:"userannounce"`
	}
	if err := json.Unmarshal(body, &out); err != nil {
		return "", "", fmt.Errorf("engine: parsing userannounce response: %w", err)
	}
	if out.UserAnnounce == "" {
		return "", "", fmt.Errorf("engine: userannounce response had no announce URL")
	}

	// Derive the ipv6.academictorrents.com variant of the returned URL, so a
	// torrent listing the IPv6 tracker gets an equivalent keyed announce
	// instead of an unkeyed fallback. The main host and the passkey stay
	// exactly as AT returned them.
	ipv6 := ""
	if u, err := url.Parse(out.UserAnnounce); err == nil && u.Hostname() == "academictorrents.com" {
		u.Host = "ipv6.academictorrents.com"
		ipv6 = u.String()
	}

	return out.UserAnnounce, ipv6, nil
}

// keyedTrackers rewrites a tiered tracker list (as found on a TorrentSpec,
// i.e. mi.UpvertedAnnounceList) so that any Academic Torrents tracker is
// replaced with the operator's per-user announce URL. Third-party trackers
// and any non-matching URL are passed through untouched. If userAnnounce is
// empty (no key configured, or resolution failed), the list is returned
// unchanged.
func keyedTrackers(trackers [][]string, userAnnounce, userAnnounceIPv6 string) [][]string {
	if userAnnounce == "" {
		return trackers
	}
	out := make([][]string, 0, len(trackers))
	for _, tier := range trackers {
		newTier := make([]string, len(tier))
		for i, announce := range tier {
			if keyed := atAnnounceURL(announce, userAnnounce, userAnnounceIPv6); keyed != "" {
				newTier[i] = keyed
			} else {
				newTier[i] = announce
			}
		}
		out = append(out, newTier)
	}
	return out
}

// atTrackersOnly filters a tiered tracker list down to Academic Torrents'
// own trackers (swapped for the operator's per-user announce URL when one
// is configured), dropping every third-party tracker. Held torrents get
// this treatment (see addTorrentSpec): the underlying client re-announces
// to every tracker in a torrent's spec on its own schedule, and AT catalog
// entries list up to a dozen mostly-dead third-party trackers each - so
// keeping them meant a node holding hundreds of torrents burned CPU
// cycling announce timeouts against dead public trackers. keep-at is an AT
// seeder; AT's tracker and DHT are its peer discovery. The network-status
// census's probe client uses the same filtering.
//
// When the list contains no AT tracker at all (a non-AT catalog entry, a
// hand-built torrent, a test stub), it's returned unchanged - filtering to
// zero trackers would leave the torrent with no tracker-based peer
// discovery whatsoever, which is worse than keeping the third-party list.
func atTrackersOnly(trackers [][]string, userAnnounce, userAnnounceIPv6 string) [][]string {
	foundAT := false
	for _, tier := range trackers {
		for _, announce := range tier {
			if atAnnounceURL(announce, userAnnounce, userAnnounceIPv6) != "" {
				foundAT = true
				break
			}
		}
		if foundAT {
			break
		}
	}
	if !foundAT {
		return trackers
	}

	out := make([][]string, 0, len(trackers))
	for _, tier := range trackers {
		newTier := make([]string, 0, len(tier))
		for _, announce := range tier {
			if keyed := atAnnounceURL(announce, userAnnounce, userAnnounceIPv6); keyed != "" {
				newTier = append(newTier, keyed)
			}
		}
		if len(newTier) > 0 {
			out = append(out, newTier)
		}
	}
	return out
}

File: src/atkey.rs
//! Academic Torrents API-key handling: resolve the operator's key to a
//! per-user announce URL, and rewrite tracker lists to use it only on AT's
//! own https hosts.
//!
//! Security contract (preserved): the key and derived URLs carry the
//! account's passkey - they are sent only to academictorrents.com tracker
//! hosts, never logged, never written to disk.

use anyhow::{Context, Result};

pub const USER_ANNOUNCE_ENDPOINT: &str = "https://academictorrents.com/apiv2/userannounce";

/// Resolve the API key (`uid=N;pass=H`, valid Cookie syntax) to the
/// per-user announce URL plus its ipv6-host variant.
pub async fn resolve_user_announce(
    client: &reqwest::Client,
    api_key: &str,
) -> Result<(String, String)> {
    let resp = client
        .get(USER_ANNOUNCE_ENDPOINT)
        .header("Cookie", api_key)
        .send()
        .await
        .context("fetching user announce URL")?;
    let status = resp.status();
    let body = resp
        .bytes()
        .await
        .context("reading userannounce response")?;
    if !status.is_success() {
        anyhow::bail!("userannounce endpoint returned {status}");
    }
    let body = &body[..body.len().min(1 << 20)];
    let out: UserAnnounce =
        serde_json::from_slice(body).context("parsing userannounce response")?;
    if out.userannounce.is_empty() {
        anyhow::bail!("userannounce response had no announce URL");
    }
    let ipv6 = upgrade_to_ipv6_host(&out.userannounce);
    Ok((out.userannounce, ipv6))
}

#[derive(serde::Deserialize)]
struct UserAnnounce {
    #[serde(default)]
    userannounce: String,
}

fn upgrade_to_ipv6_host(url: &str) -> String {
    match url::Url::parse(url) {
        Ok(mut u) if u.host_str() == Some("academictorrents.com") => {
            if u.set_host(Some("ipv6.academictorrents.com")).is_ok() {
                return u.to_string();
            }
            String::new()
        }
        _ => String::new(),
    }
}

/// The per-user announce URL to use for a tracker URL, or None when the URL
/// is not an https URL on an AT tracker host (must never receive the key).
pub fn at_announce_url(
    announce: &str,
    user_announce: &str,
    user_announce_ipv6: &str,
) -> Option<String> {
    let u = url::Url::parse(announce).ok()?;
    if u.scheme() != "https" {
        return None;
    }
    match u.host_str()? {
        "academictorrents.com" if !user_announce.is_empty() => Some(user_announce.to_string()),
        "ipv6.academictorrents.com" if !user_announce_ipv6.is_empty() => {
            Some(user_announce_ipv6.to_string())
        }
        _ => None,
    }
}

/// Rewrite tracker tiers so AT trackers use the per-user URL. Third-party
/// trackers pass through untouched. Empty key => unchanged.
pub fn keyed_trackers(
    tiers: Vec<Vec<String>>,
    user_announce: &str,
    user_announce_ipv6: &str,
) -> Vec<Vec<String>> {
    if user_announce.is_empty() {
        return tiers;
    }
    tiers
        .into_iter()
        .map(|tier| {
            tier.into_iter()
                .map(|t| at_announce_url(&t, user_announce, user_announce_ipv6).unwrap_or(t))
                .collect()
        })
        .collect()
}

/// Filter tiers down to AT's own trackers (keyed when configured), dropping
/// third-party trackers. When no AT tracker is present at all, the list is
/// returned unchanged - zero trackers would leave no tracker discovery.
pub fn at_trackers_only(
    tiers: Vec<Vec<String>>,
    user_announce: &str,
    user_announce_ipv6: &str,
) -> Vec<Vec<String>> {
    let found = tiers.iter().flatten().any(|t| {
        at_announce_url(t, user_announce, user_announce).is_some() || is_at_tracker_url(t)
    });
    if !found {
        return tiers;
    }
    tiers
        .into_iter()
        .filter_map(|tier| {
            let kept: Vec<String> = tier
                .into_iter()
                .filter_map(|t| {
                    if is_at_tracker_url(&t) {
                        Some(at_announce_url(&t, user_announce, user_announce_ipv6).unwrap_or(t))
                    } else {
                        None
                    }
                })
                .collect();
            if kept.is_empty() {
                None
            } else {
                Some(kept)
            }
        })
        .collect()
}

pub fn is_at_tracker_url(url: &str) -> bool {
    match url::Url::parse(url) {
        Ok(u) => u
            .host_str()
            .map(|h| h.contains("academictorrents.com"))
            .unwrap_or(false),
        Err(_) => false,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn keying() {
        let tiers = vec![vec![
            "https://academictorrents.com/announce.php".to_string(),
            "udp://third:80/announce".to_string(),
        ]];
        let keyed = keyed_trackers(
            tiers.clone(),
            "https://academictorrents.com/announce.php?passkey=K",
            "",
        );
        assert_eq!(
            keyed[0][0],
            "https://academictorrents.com/announce.php?passkey=K"
        );
        assert_eq!(keyed[0][1], "udp://third:80/announce");

        let only = at_trackers_only(
            tiers,
            "https://academictorrents.com/announce.php?passkey=K",
            "",
        );
        assert_eq!(only.len(), 1);
        assert_eq!(only[0].len(), 1);

        // No AT tracker: unchanged.
        let third = vec![vec!["udp://third:80/announce".to_string()]];
        assert_eq!(at_trackers_only(third.clone(), "K", ""), third);

        // http (non-https) AT URL never receives the key.
        assert_eq!(
            at_announce_url("http://academictorrents.com/announce.php", "K", ""),
            None
        );
    }
}
